community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/LuceneIndexAccessor.java

```diff
@@ -212,5 +205,87 @@ public void remove( Iterable<Long> nodeIds ) throws IOException
             }
         }
     }
+
+    private class LucenePreparedIndexUpdates implements PreparedIndexUpdates
+    {
+        final Iterable<NodePropertyUpdate> updates;
+        final int additions;
+        final boolean inRecovery;
+
+        boolean committed;
+        boolean rolledBack;
+
+        LucenePreparedIndexUpdates( Iterable<NodePropertyUpdate> updates, int additions, boolean inRecovery )
+        {
+            this.updates = updates;
+            this.additions = additions;
+            this.inRecovery = inRecovery;
+        }
+
+        @Override
+        public void commit() throws IOException
+        {
+            checkStatus();
+            try
+            {
+                committed = true;
+                for ( NodePropertyUpdate update : updates )
+                {
+                    switch ( update.getUpdateMode() )
+                    {
+                    case ADDED:
+                        if ( inRecovery )
+                        {
+                            addRecovered( update.getNodeId(), update.getValueAfter() );
+                        }
+                        else
+                        {
+                            add( update.getNodeId(), update.getValueAfter() );
+                        }
+                        break;
+                    case CHANGED:
+                        change( update.getNodeId(), update.getValueAfter() );
+                        break;
+                    case REMOVED:
+                        LuceneIndexAccessor.this.remove( update.getNodeId() );
+                        break;
+                    default:
+                        throw new UnsupportedOperationException( "Unknown update mode: " + update.getUpdateMode() );
+                    }
+                }
+
+                refreshSearcherManager(); // to allow all following reads see committed changes
+            }
+            finally
+            {
+                withdrawReservation();
+            }
+        }
+
+        @Override
+        public void rollback()
+        {
+            checkStatus();
+            rolledBack = true;
+            withdrawReservation();
+        }
+
+        void checkStatus()
+        {
+            if ( committed )
+            {
+                throw new IllegalStateException( "Index updates were already committed" );
+            }
+            if ( rolledBack )
+            {
+                throw new IllegalStateException( "Index updates were already rolled back" );
+            }
+        }
+
+        void withdrawReservation()
```

Comment:
I'm not liking "withdraw", how about calling it "releaseReservations()" instead?